### ISSUE TITLE
Server Crash: Concurrent read/write on map

### ISSUE THREAD

**Issue by kesslermaximilian** [Hanabi-Live/hanabi-live#2855]
James posted this crash log on Discord:

[message.txt](https://github.com/Hanabi-Live/hanabi-live/files/12179609/message.txt)

**Comment by kesslermaximilian** [comment]
Probably this is a bug introduced in commit e74d6bdf in the the `ActiveSpectators` method in `table.go:433`.

**Comment by Zamiell** [comment]
@Krycke

**Comment by Krycke** [comment]
Yeah, the record of spectators on the table hold a list of all spectators that have been specific since the table was created. The spectator may or may not spectate right now. 

The only trust worthy record of if a spectator is spectating actively right now is the record in tables that we get from the function `tables.GetTablesUserSpectating(sp.UserID)`. 

So now from the point of view of the table, we need to go through the list of all spectators, and see if any of them are spectating a table right now, and if they take appear to be this one.

This function in it self should not be the problem here, i have a hard time reading the stack trace on my phone, but i assume that there is a deadlock. 

When i fixed this function, i was also confused about what we are locking and when, this function it self does not need the tables to be locked, if they are changed during checking it wouldn't matter, but this function uses functions that are might or might not need a lock.

The problem with finding dead locks in the stack trace is that it's not this thread that might be the problem, it's the thread that has claimed the lock the is suck somewhere that probably is the problem. But that one is not seen in the trace.

When i was trying to investigate the dead locks before i remember that the following could occur: `thread 1` wants to first acquire `lock 1` and while having that lock, it is trying to acquire `lock 2`. The second lock is already kept by a `thread 2`, so `thread 1` gets stuck. At the same time a `thread 3` is trying to acquire `lock 1` and eventually times out. That's why we find the stack trace from the dead lock on `thread 3` but i think the problem is actually that `thread 2` won't let go of `lock 2` or that `thread 1` is trying to acquire a second lock before releasing the first one.

Either way i think all locks need to be checked in all functions, because something is strange with them.

**Comment by kesslermaximilian** [comment]
I think the function here should definitely have the lock for the tables struct when being executed. I can see why we wouldn't care if the spectators list changes while being read, but (presumably) these reads are not atomic, so stuff might go wrong if another thread writes these at the same time.
The error message from the trace explicitly says 'concurrent read and write', so it could very much be that the stack trace we see is the thread reading from the tables struct, while some other thread (hopefully holding the lock for that) is currently writing it because some spectator joined.

I also don't see how this resembles a deadlock at the moment. I don't know go much, but for a deadlock, I would expect a differen terror message, because there is actually no read or write occurring if a deadlock happens.

Also, do you know why `tables.GetTablesUserSpectating` is the only reliable source for the active spectators? If a user leaves a table, why would we not update the list at the table itself, i.e. `table.Spectators`? To me it would seem that this should be what we want to have (do we ever need these past spectators?), so probably we should just fix that and not look at the `tables` struct here at all.

Regarding possible deadlocks in the Server code: Sure, there might be some, I have no idea and it would be quite a bit of work to check them all.

**Comment by Krycke** [comment]
Ok, as i said, i could read the error message on my phone.

Before my changes that was actually how it worked, the spectator was removed when the spectator left. But the problem if that the messages written by the spectator is also supported in this struct, so by removing it we also lose what a spectator has written, that's why i changed it to active and inactive spectators. 

This was to solve at least two things: if a spectator leaves and comes back (even just reloads the page) the messages should stay. This is extra important when shadowing a game. Secondly when the game has finished, the spectators are remade and then the messages are lost.

**Comment by Krycke** [comment]
In the beginning of the redesign i was creating a separate list of active spectators, but then it could diverge from the table list. So it's better to have the information at only one place

**Comment by Zamiell** [comment]
> but i assume that there is a deadlock.

To be clear, there is not a deadlock. Historically, when the server crashes, it is due to a deadlock, but not this time. This time, the error message is this:

> fatal error: concurrent map read and map write

This means that a two goroutines are attempting to read and write to a map at the same time.
Maps in Golang are not safe for concurrent use, so you are supposed to lock the corresponding map mutex first before doing any reads/writes.

**Comment by Zamiell** [comment]
@Krycke do you have time to try and fix this this weekend? i dont want the server to crash again

**Comment by Krycke** [comment]
Sorry, i don't have time coding at the moment, I will probably not sit in front of a computer for the foreseeable future.

**Comment by Zamiell** [comment]
I'll revert the PR then

**Comment by Zamiell** [comment]
Unless @kesslermaximilian wants to take a look?
